File: generate_chat_html.py
# generate_chat_html.py
# to run: python3 generate_chat_html.py

import os
import re
from datetime import datetime
import git
import chardet

def read_file(file_path):
	with open(file_path, 'r') as file:
		return file.read()

def extract_metadata(content):
	author = re.search(r'Author:\s*(.+)', content, re.IGNORECASE)
	author = author.group(1) if author else "Unknown"
	hashtags = re.findall(r'#\w+', content)
	return author, hashtags

def truncate_message(content, max_length=300):
	if len(content) <= max_length:
		return content, False
	return content[:max_length] + "...", True

import os
from os import scandir
import git
from datetime import datetime
import chardet

def generate_chat_html(repo_path, output_file):
	repo = git.Repo(repo_path)
	HTML_TEMPLATE = read_file('./template/html/chat_page.html')
	MESSAGE_TEMPLATE = read_file('./template/html/chat_message.html')
	CSS_STYLE = read_file('./template/css/chat_style.css')
	JS_TEMPLATE = read_file('./template/js/chat.js')

	messages = []
	message_dir = os.path.join(repo_path, "message")

	def scan_directory(directory):
		with scandir(directory) as entries:
			for entry in entries:
				if entry.is_file() and entry.name.endswith(".txt"):
					process_file(entry.path)
				elif entry.is_dir():
					scan_directory(entry.path)

	def process_file(file_path):
		relative_path = os.path.relpath(file_path, repo_path)

		try:
			commit = next(repo.iter_commits(paths=relative_path, max_count=1))
			commit_timestamp = datetime.fromtimestamp(commit.committed_date)
		except StopIteration:
			commit_timestamp = datetime.min

		try:
			with open(file_path, 'rb') as file:
				raw_data = file.read()
			detected_encoding = chardet.detect(raw_data)['encoding']
			with open(file_path, 'r', encoding=detected_encoding, errors='ignore') as file:
				content = file.read()
			author, hashtags = extract_metadata(content)
		except Exception as e:
			print(f"Error reading file {file_path}: {str(e)}")
			author = "Error"
			content = "Error reading message"
			hashtags = []

		messages.append({
			'author': author,
			'content': content,
			'timestamp': commit_timestamp,
			'hashtags': hashtags
		})

	scan_directory(message_dir)

	# Sort messages by timestamp in ascending order
	messages.sort(key=lambda x: x['timestamp'])

	chat_messages = []
	for idx, msg in enumerate(messages):
		truncated_content, is_truncated = truncate_message(msg['content'])
		expand_link = f'<a href="#" class="expand-link" data-message-id="{idx}">{"Show More" if is_truncated else "Show Less"}</a>'
		full_content = f'<div class="full-message" id="full-message-{idx}" style="display: none;">{msg["content"]}</div>' if is_truncated else ''

		chat_messages.append(MESSAGE_TEMPLATE.format(
			author=msg['author'],
			content=truncated_content,
			full_content=full_content,
			expand_link=expand_link,
			timestamp=msg['timestamp'].strftime('%Y-%m-%d %H:%M:%S'),
			hashtags=' '.join(msg['hashtags'])
		))

	html_content = HTML_TEMPLATE.format(
		style=CSS_STYLE,
		chat_messages=''.join(chat_messages),
		message_count=len(messages),
		current_time=datetime.now().strftime('%Y-%m-%d %H:%M:%S'),
		title="THIMBLE Chat"
	)

	html_content = html_content.replace('</body>', '<script>' + JS_TEMPLATE + '</script>' + '</body>')

	with open(output_file, 'w', encoding='utf-8') as f:
		f.write(html_content)

if __name__ == "__main__":
	repo_path = "."  # Current directory
	output_file = 'chat.html'
	generate_chat_html(repo_path, output_file)
	print(f"Chat log generated: {output_file}")

File: generate_chat_html.js
// generate_chat_html.js
// to run: node generate_chat_html.js

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');
const chardet = require('chardet');
const argparse = require('argparse');

// Debug flag
let DEBUG = false;

function debugPrint(...args) {
    if (DEBUG) {
        console.log(...args);
    }
}

debugPrint("Script started.");

function readFile(filePath) {
    debugPrint(`Reading file: ${filePath}`);
    const content = fs.readFileSync(filePath, 'utf8');
    debugPrint(`File read successfully. Content length: ${content.length} characters`);
    return content;
}

function extractMetadata(content) {
    debugPrint("Extracting metadata from content");
    const authorMatch = content.match(/Author:\s*(.+)/i);
    const author = authorMatch ? authorMatch[1] : "Unknown";
    const hashtags = (content.match(/#\w+/g) || []);
    debugPrint(`Extracted metadata - Author: ${author}, Hashtags: ${hashtags}`);
    return [author, hashtags];
}

function truncateMessage(content, maxLength = 300) {
    debugPrint(`Truncating message. Original length: ${content.length}`);
    if (content.length <= maxLength) {
        debugPrint("Message does not need truncation");
        return [content, false];
    }
    const truncated = content.slice(0, maxLength) + "...";
    debugPrint(`Message truncated. New length: ${truncated.length}`);
    return [truncated, true];
}

function generateChatHtml(repoPath, outputFile, maxMessages = 50, maxMessageLength = 300, title = "THIMBLE Chat") {
    debugPrint(`Generating chat HTML. Repo path: ${repoPath}, Output file: ${outputFile}`);
    const HTML_TEMPLATE = readFile('./template/html/chat_page.html');
    const MESSAGE_TEMPLATE = readFile('./template/html/chat_message.html');
    const CSS_STYLE = readFile('./template/css/chat_style.css');
    const JS_TEMPLATE = readFile('./template/js/chat.js');

    const messages = [];
    const messageDir = path.join(repoPath, "message");
    debugPrint(`Scanning directory: ${messageDir}`);

    function scanDirectory(directory) {
        debugPrint(`Scanning directory: ${directory}`);
        const entries = fs.readdirSync(directory, { withFileTypes: true });
        for (const entry of entries) {
            if (entry.isFile() && entry.name.endsWith(".txt")) {
                debugPrint(`Processing file: ${path.join(directory, entry.name)}`);
                processFile(path.join(directory, entry.name));
            } else if (entry.isDirectory()) {
                scanDirectory(path.join(directory, entry.name));
            }
        }
    }

    function processFile(filePath) {
        debugPrint(`Processing file: ${filePath}`);
        const relativePath = path.relative(repoPath, filePath);

        let commitTimestamp;
        try {
            const gitLog = execSync(`git log -1 --format=%ct -- "${relativePath}"`, { cwd: repoPath }).toString().trim();
            commitTimestamp = new Date(parseInt(gitLog) * 1000);
            debugPrint(`Commit timestamp: ${commitTimestamp}`);
        } catch (error) {
            debugPrint("No commit found for file, using minimum date");
            commitTimestamp = new Date(0);
        }

        try {
            const rawData = fs.readFileSync(filePath);
            const detectedEncoding = chardet.detect(rawData);
            debugPrint(`Detected encoding: ${detectedEncoding}`);
            const content = fs.readFileSync(filePath, detectedEncoding);
            const [author, hashtags] = extractMetadata(content);

            const cleanedContent = content.replace(/author:\s*.+/i, '').trim();
            debugPrint(`Processed content. Length: ${cleanedContent.length}`);

            messages.push({
                author,
                content: cleanedContent,
                timestamp: commitTimestamp,
                hashtags
            });
            debugPrint(`Message added. Total messages: ${messages.length}`);
        } catch (error) {
            debugPrint(`Error reading file ${filePath}: ${error.message}`);
            messages.push({
                author: "Error",
                content: "Error reading message",
                timestamp: commitTimestamp,
                hashtags: []
            });
        }
    }

    scanDirectory(messageDir);

    debugPrint("Sorting messages by timestamp");
    messages.sort((a, b) => b.timestamp - a.timestamp);

    // Limit the number of messages
    messages.splice(maxMessages);

    const chatMessages = messages.map((msg, idx) => {
        debugPrint(`Processing message ${idx + 1}/${messages.length}`);
        const [truncatedContent, isTruncated] = truncateMessage(msg.content, maxMessageLength);
        const expandLink = `<a href="#" class="expand-link" data-message-id="${idx}">${isTruncated ? "Show More" : ""}</a>`;
        const fullContent = isTruncated ? `<div class="full-message" id="full-message-${idx}" style="display: none;">${msg.content}</div>` : '';

        return MESSAGE_TEMPLATE
            .replace('{author}', msg.author)
            .replace('{content}', truncatedContent)
            .replace('{full_content}', fullContent)
            .replace('{expand_link}', expandLink)
            .replace('{timestamp}', msg.timestamp.toISOString().replace('T', ' ').substr(0, 19))
            .replace('{hashtags}', msg.hashtags.join(' '));
    });

    debugPrint("Generating final HTML content");
    let htmlContent = HTML_TEMPLATE
        .replace('{style}', CSS_STYLE)
        .replace('{chat_messages}', chatMessages.join(''))
        .replace('{message_count}', messages.length)
        .replace('{current_time}', new Date().toISOString().replace('T', ' ').substr(0, 19))
        .replace('{title}', title);

    htmlContent = htmlContent.replace('</body>', `<script>${JS_TEMPLATE}</script></body>`);

    debugPrint(`Writing HTML content to file: ${outputFile}`);
    fs.writeFileSync(outputFile, htmlContent, 'utf8');
}

if (require.main === module) {
    const parser = new argparse.ArgumentParser({
        description: "Generate chat HTML from repository messages."
    });
    parser.add_argument("--repo_path", { default: ".", help: "Path to the repository" });
    parser.add_argument("--output_file", { default: "chat.html", help: "Output HTML file name" });
    parser.add_argument("--max_messages", { type: 'int', default: 50, help: "Maximum number of messages to display" });
    parser.add_argument("--max_message_length", { type: 'int', default: 300, help: "Maximum length of each message before truncation" });
    parser.add_argument("--title", { default: "THIMBLE Chat", help: "Title of the chat page" });
    parser.add_argument("--debug", { action: "store_true", help: "Enable debug output" });

    const args = parser.parse_args();

    // Set the debug flag based on the command-line argument
    DEBUG = args.debug;

    generateChatHtml(args.repo_path, args.output_file, args.max_messages, args.max_message_length, args.title);
    debugPrint(`Chat log generated: ${args.output_file}`);
    debugPrint("Script completed.");
}
